File: pharma_book/scripts/migrate_sqlite_to_supabase.py
#!/usr/bin/env python3
"""
SQLite -> Supabase migration utility for Cyber Pharma.

Features
- Reads from local SQLite `app.db`.
- For each known table, exports to CSV under ../exports/ OR uploads to Supabase directly.
- Adds `pharmacy_id` (UUID) to tenant tables with a provided constant.
- Skips unused columns from `alt_rates` beyond {ndc,wac,pkg_size,pkg_size_mult,generic_indicator}.

Usage
  MODE=csv python scripts/migrate_sqlite_to_supabase.py
  MODE=supabase SUPABASE_URL=... SUPABASE_SERVICE_ROLE=... PHARMACY_ID=... python scripts/migrate_sqlite_to_supabase.py

Config via env vars
  MODE: 'csv' (default) or 'supabase'
  SQLITE_PATH: path to app.db (default: ../app.db)
  OUT_DIR: export directory when MODE=csv (default: ../exports)
  SUPABASE_URL, SUPABASE_SERVICE_ROLE: required when MODE=supabase
  PHARMACY_ID: UUID to stamp on tenant tables (if not set, a new UUID is generated and printed)

Notes
- When MODE=supabase, this uses the Supabase Python client. Ensure package 'supabase' v2+ is installed.
- Import order respects reference tables first, then tenant tables.
- You should have already executed README_DDL.sql in your Supabase project.
"""
from __future__ import annotations
import csv
import os
import sqlite3
import uuid
from datetime import datetime
from typing import List, Dict, Any

MODE = os.getenv('MODE', 'csv').lower()
SQLITE_PATH = os.getenv('SQLITE_PATH', os.path.join(os.path.dirname(os.path.dirname(__file__)), 'app.db'))
OUT_DIR = os.getenv('OUT_DIR', os.path.join(os.path.dirname(os.path.dirname(__file__)), 'exports'))
SUPABASE_URL = os.getenv('SUPABASE_URL')
SUPABASE_SERVICE_ROLE = os.getenv('SUPABASE_SERVICE_ROLE')
PHARMACY_ID = os.getenv('PHARMACY_ID')

TENANT_TABLES = {
    'user_data': 'pharma_user_data',
    'report_files': 'pharma_report_files',
    'pharmacy_profile': 'pharma_pharmacy_profile',
}
REF_TABLES = {
    'baseline': 'pharma_baseline',
    'alt_rates': 'pharma_alt_rates',
    'pbm_info': 'pharma_pbm_info',
}

EXPORT_COLUMNS = {
    'pharma_baseline': ['ndc','drug_name','bg','effective_date','aac'],
    'pharma_alt_rates': ['ndc','wac','pkg_size','pkg_size_mult','generic_indicator'],
    'pharma_pbm_info': ['bin','pbm_name','email'],
    'pharma_pharmacy_profile': ['pharmacy_id','pharmacy_name','address','phone','fax','email','ncpdp','npi','contact_person','created_at','updated_at'],
    'pharma_user_data': ['script','pharmacy_id','date_dispensed','drug_ndc','drug_name','qty','total_paid','new_paid','bin','pdf_file','status','created_at','updated_at'],
    'pharma_report_files': ['script','report_type','pharmacy_id','pdf_file','created_at'],
}

NOW = datetime.utcnow().isoformat()


def get_conn(path: str):
    conn = sqlite3.connect(path)
    conn.row_factory = sqlite3.Row
    return conn


def ensure_dir(p: str):
    os.makedirs(p, exist_ok=True)


def coerce_decimal(v):
    if v is None:
        return None
    try:
        return float(v)
    except Exception:
        return None


def export_csv(table: str, rows: List[Dict[str, Any]]):
    ensure_dir(OUT_DIR)
    path = os.path.join(OUT_DIR, f"{table}.csv")
    cols = EXPORT_COLUMNS[table]
    with open(path, 'w', newline='') as f:
        w = csv.DictWriter(f, fieldnames=cols)
        w.writeheader()
        for r in rows:
            w.writerow({c: r.get(c, '') for c in cols})
    print(f"wrote {len(rows)} rows -> {path}")


def upsert_supabase(table: str, rows: List[Dict[str, Any]]):
    from supabase import create_client, Client  # type: ignore
    supabase: Client = create_client(SUPABASE_URL, SUPABASE_SERVICE_ROLE)
    # chunk insert to avoid payload limits
    BATCH = 1000
    for i in range(0, len(rows), BATCH):
        chunk = rows[i:i+BATCH]
        resp = supabase.table(table).upsert(chunk, on_conflict="script" if table=='pharma_user_data' else None).execute()
        if getattr(resp, 'error', None):
            raise RuntimeError(resp.error)
    print(f"upserted {len(rows)} rows -> {table}")


def fetch_all(cur, sql: str, params=()):
    cur.execute(sql, params)
    return [dict(r) for r in cur.fetchall()]


def main():
    pharmacy_id = uuid.UUID(PHARMACY_ID) if PHARMACY_ID else uuid.uuid4()
    if not PHARMACY_ID:
        print(f"PHARMACY_ID not provided. Using generated: {pharmacy_id}")

    conn = get_conn(SQLITE_PATH)
    cur = conn.cursor()

    # Reference tables
    out_rows = {}
    out_rows['pharma_baseline'] = [
        {
            'ndc': r.get('ndc') or r.get('drug_ndc') or r.get('NDC') or r.get('Ndc'),
            'drug_name': r.get('drug name') or r.get('drug_name') or r.get('DrugName'),
            'bg': r.get('bg') or r.get('BG'),
            'effective_date': r.get('effective date') or r.get('effective_date'),
            'aac': r.get('aac')
        }
        for r in fetch_all(cur, 'select * from baseline')
    ]

    out_rows['pharma_alt_rates'] = [
        {
            'ndc': r.get('ndc'),
            'wac': coerce_decimal(r.get('wac')),
            'pkg_size': coerce_decimal(r.get('pkg_size')),
            'pkg_size_mult': coerce_decimal(r.get('pkg_size_mult')),
            'generic_indicator': r.get('generic_indicator') or r.get('generic') or r.get('generic_flag')
        }
        for r in fetch_all(cur, 'select * from alt_rates')
    ]

    out_rows['pharma_pbm_info'] = [
        {'bin': r.get('bin'), 'pbm_name': r.get('pbm_name'), 'email': r.get('email')}
        for r in fetch_all(cur, 'select * from pbm_info')
    ]

    # Tenant tables
    # pharmacy_profile (single row expected)
    prof = fetch_all(cur, 'select * from pharmacy_profile limit 1')
    if prof:
        p = prof[0]
        out_rows['pharma_pharmacy_profile'] = [{
            'pharmacy_id': str(pharmacy_id),
            'pharmacy_name': p.get('pharmacy_name'),
            'address': p.get('address'),
            'phone': p.get('phone'),
            'fax': p.get('fax'),
            'email': p.get('email'),
            'ncpdp': p.get('ncpdp'),
            'npi': p.get('npi'),
            'contact_person': p.get('contact_person'),
            'created_at': NOW,
            'updated_at': NOW,
        }]
    else:
        out_rows['pharma_pharmacy_profile'] = [{
            'pharmacy_id': str(pharmacy_id),
            'pharmacy_name': 'Unknown Pharmacy',
            'address': '', 'phone': '', 'fax': '', 'email': '',
            'ncpdp': '', 'npi': '', 'contact_person': '',
            'created_at': NOW, 'updated_at': NOW
        }]

    # user_data
    out_rows['pharma_user_data'] = []
    for r in fetch_all(cur, 'select * from user_data'):
        out_rows['pharma_user_data'].append({
            'script': r.get('script'),
            'pharmacy_id': str(pharmacy_id),
            'date_dispensed': r.get('date_dispensed'),
            'drug_ndc': r.get('drug_ndc') or r.get('ndc'),
            'drug_name': r.get('drug_name'),
            'qty': r.get('qty'),
            'total_paid': r.get('total_paid'),
            'new_paid': r.get('new_paid'),
            'bin': r.get('bin'),
            'pdf_file': r.get('pdf_file'),
            'status': r.get('status'),
            'created_at': NOW,
            'updated_at': NOW,
        })

    # report_files
    out_rows['pharma_report_files'] = []
    for r in fetch_all(cur, 'select * from report_files'):
        out_rows['pharma_report_files'].append({
            'script': r.get('script'),
            'report_type': r.get('report_type'),
            'pharmacy_id': str(pharmacy_id),
            'pdf_file': r.get('pdf_file'),
            'created_at': NOW,
        })

    # Export or upload
    if MODE == 'csv':
        for table, rows in out_rows.items():
            export_csv(table, rows)
        print("CSV exports complete →", OUT_DIR)
    elif MODE == 'supabase':
        missing = [k for k in ('SUPABASE_URL','SUPABASE_SERVICE_ROLE') if not globals().get(k)]
        if missing:
            raise SystemExit(f"Missing env vars: {missing}")
        # Load order: references then tenant tables
        for t in ('pharma_baseline','pharma_alt_rates','pharma_pbm_info','pharma_pharmacy_profile','pharma_user_data','pharma_report_files'):
            upsert_supabase(t, out_rows[t])
        print("Supabase upload complete.")
    else:
        raise SystemExit("MODE must be 'csv' or 'supabase'")

if __name__ == '__main__':
    main()

File: pharma_book/scripts/migrateSqliteToSupabase.js
#!/usr/bin/env node
"use strict";
/**
 * SQLite -> Supabase migration for Cyber Pharma (Node.js).
 * MODE:
 *   - csv       : export CSVs to OUT_DIR (default ./exports)
 *   - supabase  : upsert rows into Supabase using service role key
 *
 * Env:
 *   MODE=csv|supabase
 *   SQLITE_PATH=./app.db
 *   OUT_DIR=./exports
 *   NEXT_PUBLIC_SUPABASE_URL=...
 *   SUPABASE_SERVICE_ROLE_KEY=...
 *   PHARMACY_ID=<uuid>  (optional; generated fallback)
 */
const fs = require('fs');
const path = require('path');
const Database = require('better-sqlite3');
const { createClient } = require('@supabase/supabase-js');

const MODE = (process.env.MODE || 'csv').toLowerCase();
const SQLITE_PATH = process.env.SQLITE_PATH || './app.db';
const OUT_DIR = process.env.OUT_DIR || './exports';
const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL;
const SERVICE_ROLE = process.env.SUPABASE_SERVICE_ROLE_KEY;

const NOW = new Date().toISOString();
const PHARMACY_ID = (process.env.PHARMACY_ID && /^[0-9a-f-]{36}$/i.test(process.env.PHARMACY_ID))
  ? process.env.PHARMACY_ID
  : cryptoUuid();

function cryptoUuid() {
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, c => {
    const r = Math.random()*16|0, v = c === 'x' ? r : (r&0x3|0x8);
    return v.toString(16);
  });
}

function writeCsv(fp, header, rows) {
  fs.mkdirSync(path.dirname(fp), { recursive: true });
  const esc = (v) => {
    if (v === null || v === undefined) return '';
    const s = String(v);
    return /[",\n]/.test(s) ? `"${s.replace(/\"/g, '""')}"` : s;
  };
  const out = [header, ...rows].map(r => r.map(esc).join(',')).join('\n') + '\n';
  fs.writeFileSync(fp, out);
  console.log(`wrote ${rows.length} rows -> ${fp}`);
}

function num(v) {
  if (v === null || v === undefined) return null;
  const n = Number(v);
  return Number.isFinite(n) ? n : null;
}

async function main() {
  const db = new Database(SQLITE_PATH, { fileMustExist: true });
  console.log(`Using PHARMACY_ID=${PHARMACY_ID}`);

  // Reference tables
  const baseline = db.prepare('select * from baseline').all().map(r => ({
    ndc: r.ndc || r.drug_ndc || r.NDC || r.Ndc,
    drug_name: r['drug name'] ?? r.drug_name ?? r.DrugName ?? '',
    bg: r.bg ?? r.BG ?? '',
    effective_date: r['effective date'] ?? r.effective_date ?? null,
    aac: r.aac ?? null
  }));

  const altRates = db.prepare('select * from alt_rates').all().map(r => ({
    ndc: r.ndc,
    wac: num(r.wac),
    pkg_size: num(r.pkg_size),
    pkg_size_mult: num(r.pkg_size_mult),
    generic_indicator: r.generic_indicator ?? r.generic ?? r.generic_flag ?? null
  }));

  const pbmInfo = db.prepare('select * from pbm_info').all().map(r => ({
    bin: r.bin, pbm_name: r.pbm_name, email: r.email
  }));

  // Tenant tables
  const profRow = db.prepare('select * from pharmacy_profile limit 1').get();
  const profile = [{
    pharmacy_id: PHARMACY_ID,
    pharmacy_name: profRow?.pharmacy_name || 'Unknown Pharmacy',
    address: profRow?.address || '',
    phone: profRow?.phone || '',
    fax: profRow?.fax || '',
    email: profRow?.email || '',
    ncpdp: profRow?.ncpdp || '',
    npi: profRow?.npi || '',
    contact_person: profRow?.contact_person || '',
    created_at: NOW,
    updated_at: NOW
  }];

  const userData = db.prepare('select * from user_data').all().map(r => ({
    script: r.script,
    pharmacy_id: PHARMACY_ID,
    date_dispensed: r.date_dispensed,
    drug_ndc: r.drug_ndc || r.ndc || null,
    drug_name: r.drug_name || null,
    qty: r.qty,
    total_paid: r.total_paid,
    new_paid: r.new_paid,
    bin: r.bin,
    pdf_file: r.pdf_file,
    status: r.status,
    created_at: NOW,
    updated_at: NOW
  }));

  const reportFiles = db.prepare('select * from report_files').all().map(r => ({
    script: r.script,
    report_type: r.report_type,
    pharmacy_id: PHARMACY_ID,
    pdf_file: r.pdf_file,
    created_at: NOW
  }));

  if (MODE === 'csv') {
    writeCsv(path.join(OUT_DIR, 'pharma_baseline.csv'),
      ['ndc','drug_name','bg','effective_date','aac'],
      baseline.map(r => [r.ndc, r.drug_name, r.bg, r.effective_date, r.aac]));

    writeCsv(path.join(OUT_DIR, 'pharma_alt_rates.csv'),
      ['ndc','wac','pkg_size','pkg_size_mult','generic_indicator'],
      altRates.map(r => [r.ndc, r.wac, r.pkg_size, r.pkg_size_mult, r.generic_indicator]));

    writeCsv(path.join(OUT_DIR, 'pharma_pbm_info.csv'),
      ['bin','pbm_name','email'],
      pbmInfo.map(r => [r.bin, r.pbm_name, r.email]));

    writeCsv(path.join(OUT_DIR, 'pharma_pharmacy_profile.csv'),
      ['pharmacy_id','pharmacy_name','address','phone','fax','email','ncpdp','npi','contact_person','created_at','updated_at'],
      profile.map(r => [r.pharmacy_id,r.pharmacy_name,r.address,r.phone,r.fax,r.email,r.ncpdp,r.npi,r.contact_person,r.created_at,r.updated_at]));

    writeCsv(path.join(OUT_DIR, 'pharma_user_data.csv'),
      ['script','pharmacy_id','date_dispensed','drug_ndc','drug_name','qty','total_paid','new_paid','bin','pdf_file','status','created_at','updated_at'],
      userData.map(r => [r.script,r.pharmacy_id,r.date_dispensed,r.drug_ndc,r.drug_name,r.qty,r.total_paid,r.new_paid,r.bin,r.pdf_file,r.status,r.created_at,r.updated_at]));

    writeCsv(path.join(OUT_DIR, 'pharma_report_files.csv'),
      ['script','report_type','pharmacy_id','pdf_file','created_at'],
      reportFiles.map(r => [r.script,r.report_type,r.pharmacy_id,r.pdf_file,r.created_at]));

    console.log('CSV exports complete ->', OUT_DIR);
  } else if (MODE === 'supabase') {
    if (!SUPABASE_URL || !SERVICE_ROLE) {
      throw new Error('SUPABASE credentials missing. Set NEXT_PUBLIC_SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY');
    }
    const supabase = createClient(SUPABASE_URL, SERVICE_ROLE, { auth: { persistSession: false } });

    const upsert = async (table, rows, onConflict) => {
      for (let i = 0; i < rows.length; i += 1000) {
        const chunk = rows.slice(i, i + 1000);
        const { error } = await supabase.from(table).upsert(chunk, { onConflict });
        if (error) throw error;
      }
      console.log(`upserted ${rows.length} -> ${table}`);
    };

    await upsert('pharma_baseline', baseline, 'ndc');
    await upsert('pharma_alt_rates', altRates, 'ndc');
    await upsert('pharma_pbm_info', pbmInfo, 'bin');
    await upsert('pharma_pharmacy_profile', profile, 'pharmacy_id');
    await upsert('pharma_user_data', userData, 'script');
    await upsert('pharma_report_files', reportFiles, 'script,report_type');

    console.log('Supabase upload complete.');
  } else {
    throw new Error("MODE must be 'csv' or 'supabase'");
  }
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});
